hazelcast/src/test/java/com/hazelcast/executor/ExecutorServiceTest.java

```diff
@@ -80,34 +69,187 @@
+        final ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+        try {
+            final Future<String> future = executorService.submit(new CountDownLatchAwaitingCallable(callableLatch));
+            final CountingDownExecutionCallback<String> callback = new CountingDownExecutionCallback<String>(1);
+            final ICompletableFuture<String> completableFuture = es.asCompletableFuture(future);
+            completableFuture.andThen(callback);
+
+            callableLatch.countDown();
+            assertTrue(callback.getLatch().await(30, TimeUnit.SECONDS));
+            assertEquals(CountDownLatchAwaitingCallable.RESULT, callback.getResult());
+        } finally {
+            executorService.shutdown();
+        }
+    }
+
+    @Test
+    public void testMultiPreregisteredExecutionCallbackCompletableFuture() throws Exception {
+        final ExecutionService es = getExecutionService(createHazelcastInstance());
+
+        final CountDownLatch callableLatch = new CountDownLatch(1);
+        final CountDownLatch callbackLatch = new CountDownLatch(2);
+        final ExecutorService executorService = Executors.newSingleThreadExecutor();
+        try {
+            final Future<String> future = executorService.submit(new CountDownLatchAwaitingCallable(callableLatch));
+            final CountingDownExecutionCallback<String> callback1 = new CountingDownExecutionCallback<String>(callableLatch);
+            final CountingDownExecutionCallback<String> callback2 = new CountingDownExecutionCallback<String>(callableLatch);
+
+            final ICompletableFuture<String> completableFuture = es.asCompletableFuture(future);
+            completableFuture.andThen(callback1);
+            completableFuture.andThen(callback2);
+
+            callableLatch.countDown();
+            callbackLatch.await(30, TimeUnit.SECONDS);
+            assertEquals(CountDownLatchAwaitingCallable.RESULT, callback1.getResult());
+            assertEquals(CountDownLatchAwaitingCallable.RESULT, callback2.getResult());
+        } finally {
+            executorService.shutdown();
+        }
+    }
+
+    @Test
+    public void testPostregisteredExecutionCallbackCompletableFuture() throws Exception {
+        final ExecutionService es = getExecutionService(createHazelcastInstance());
+
+        final CountDownLatch callableLatch = new CountDownLatch(1);
+        final ExecutorService executorService = Executors.newSingleThreadExecutor();
+        try {
+            final Future<String> future = executorService.submit(new CountDownLatchAwaitingCallable(callableLatch));
+            final ICompletableFuture<String> completableFuture = es.asCompletableFuture(future);
+            callableLatch.countDown();
+            future.get();
+
+            final CountingDownExecutionCallback<String> callback = new CountingDownExecutionCallback<String>(1);
+            completableFuture.andThen(callback);
+
+            try {
```

Comment:
Normally the assert message should contain the failure message. Now the assert error is caught and then the logging is done. This isn't very pretty. I would really suggest moving it back into the assert message.